### PR TITLE
flatten inputs in tf.layers.dense

### PR DESCRIPTION
**Pull request by mtjuney**
In the document of `tf.layers.dense` is writen that: 

> Note: if the inputs tensor has a rank greater than 2, then it is flattened prior to the initial matrix multiply by kernel.

but inputs is not flattened.

I modified that to flatten inputs.

### CODE REVIEW

**Comment by drpngx** [comment]
@fchollet could you take a look?

**Comment by ebrevdo** [comment]
Not sure if that the original docstring is correct.  Nor that this PR fixes
anything; more likely that it restricts the behavior of tf.layers.dense.

the current behavior is:

input is [a, b, c, ..., m], and num_units = 10

output is: [a, b, c, ..., 10]

(here a, b, c, ..., need not be known at graph build time)

this PR seems to break that guarantee.

**Comment by fchollet** [comment]
Eugene is right, the goal with the `Dense` layer is that we apply a densely-connected layer on the last axis of the input. This allows e.g. time-distributed densely-connected transformations, where the same transformation is applied at every timestep of an input. 

So we don't want to flatten the output. And besides this overarching conceptual issue, there are also many technical problems with this PR.

If the current docstring is ambiguous, then we should update it to make the layer behavior clearer. But the behavior as currently implemented is the proper one.

**Comment by joyousrabbit** [comment]
Hi, all, I searched everywhere but cannot find the difference between ft.layers.dense and tf.contrib.layers.fully_connected. 

Could you please point me out their difference?

And if dense only applies to the last axis, how can we do a real full connections of elements from all dims like in Cafffe? In Caffe, input is [N, b, c, ..., m], and num_units = 10, then output is: [N,10]

**Comment by fchollet** [comment]
> And if dense only applies to the last axis, how can we do a real full connections of elements from all dims like in Cafffe?

You could either flatten your input then apply a `Dense` layer, or write a custom layer that does the same.

> Could you please point me out their difference?

They are meant to be the same. They do have a slightly different API and a different implementation. In code going forward, you should use `tf.layers` rather than `tf.contrib.layers`, since `tf.layers` is guaranteed to be API-stable (as part of the 1.0 API) and to stay there for the long term.

**Comment by joyousrabbit** [comment]
@fchollet Thanks. Can we add a parameter to Dense (full_connected) layer to make it caffe mode for all axis but the first one? You see it's a very often meeted feature as the last layer.

**Comment by fchollet** [comment]
@joyousrabbit it's really simple to just flatten your inputs... 

In Keras:

```python
x = keras.layers.Flatten()(x)
x = keras.layers.Dense(units)(x)
```

**Comment by joyousrabbit** [comment]
@fchollet Yes, I see in tensorflow, there is a same layer do the flatten. Thanks a lot.
tf.contrib.layers.flatten
